### ISSUE TITLE
 Parse multiple yaml files at the same time

### ISSUE THREAD

**Issue by liqiyun** [AIDASoft/podio#168]
As described above, does podio support parsing multiple files at the same time ? Or say, can a yaml file be included by another yaml file ?

**Comment by tmadlener** [comment]
Hi @liqiyun,

At the moment podio only supports parsing one yaml file, which cannot include other yaml files.
In principle such a use case could be solved, by first merging all the different yaml files into one and then process that with podio.

What would be your use case for which parsing multiple yaml files would be necessary?

**Comment by liqiyun** [comment]
It will be convenient to switch between different design scheme of data models.

**Comment by SanghyunKo** [comment]
Hi @liqiyun,

I was trying similar attempts, with a slightly different motivation from yours - in my case what I was trying to do is adding several additional event data to already existing (compiled) Podio data model.

I assume feeding multiple yaml files would be possible even without 'concatenating' yaml files into a single one if you simply build Podio data models into separated CMake package and link them to your application simultaneously. 

In the case, what you need to take care is package namespace and including headers properly (if either of data model is not independent each other), since the first argument of `PODIO_GENERATE_DATAMODEL` will be the namespace, include directory name and so on - so you need to include headers by yourself if you build your model into separated packages. 
```
ExtraCode:
  includes:
    "#include <namespace/header1.h>\n #include <namespace/header2.h>"
```
But I'm not sure whether authors would encourage/discourage this kind of use case...

**Comment by tmadlener** [comment]
Hi @SanghyunKo,

That is definitely an interesting approach, thanks for pointing it out. I think this should be a supported use case, but maybe there are some things that still could be done to refine this.

You say that you put the second datamodel yaml into a separate CMake package, but I don't think that is strictly necessary. Have you tried to generate the second datamodel in the same CMake package and then simply link it against the first? Something along the lines of:
```cmake
# main datamodel
PODIO_GENERATE_DATAMODEL(datamodel1 datamodel1.yaml dm1_headers dm1_sources)
PODIO_ADD_DATAMODEL_CORE_LIB(datamodel1 "${dm1_headers}" "${dm1_sources}")
# similar for the ROOT dicts and SIO blocks if necessary

# second datamodel using parts of the first one. Generation should be unaffected
PODIO_GENERATE_DATAMODEL(datamodel2 datamodel2.yaml dm2_headers dm2_sources)
PODIO_ADD_DATAMODEL_CORE_LIB(datamodel2 "${dm2_headers}" "${dm2_sources}")
# link the main model to the secondary one
target_link_libraries(datamodel2 datamodel1)
# again, ROOT dicts, and SIO blocks for the second datamodel
```

In the yaml file of the second datamodel you still have to have the necessary extra includes as you describe it, but you could have two yaml files in one CMake package. Now of course if the first one already lives in a separate CMake package, you could just link against that.

There are currently two main limitations that I see.
- The generated source code has to be put into two different directories, because `PODIO_GENERATE_DATAMODEL` puts all the headers and source files in the directory passed to it into the returned lists.
- You have to take care of the correct extra includes in the yaml definition of your second datamodel.

Have you experienced other problems?

**Comment by SanghyunKo** [comment]
Hi @tmadlener, thanks for the prompt reply. Since what I did is adding a single yaml file of my own and linking it with another Podio model already deployed in `cvmfs` (i.e. `EDM4hep`), I've never tried to build 2 yaml files in a single package. But in principle, I would guess it should be possible except that when doing `PODIO_GENERATE_DATAMODEL` for 2nd data model the argument should be `datamodel1` rather than `datamodel2`, since we don't have direct control to the namespace and include folder name - otherwise two data models will live in different namespaces, and they should be independent in the case.

Anyway, as far as the link names (1st argument to `PODIO_ADD_DATAMODEL_CORE_LIB`) are different, it would work.

However, in the case, `podio_class_generator.py` will generate all headers and sources in the same directory as you mentioned, so the user needs to be really careful e.g. avoid declaring the same name of class in the different yaml file, or take care of directories manually.

As far as I can tell from my experience, linking two Podio data models from distinct CMake package compiles well (one is dependent to the other). I'm not sure whether there is run time issue or not, but I believe it should run fine since they're plain c++ codes after the generation :laughing:

**Comment by tmadlener** [comment]
> But in principle, I would guess it should be possible except that when doing PODIO_GENERATE_DATAMODEL for 2nd data model the argument should be datamodel1 rather than datamodel2, since we don't have direct control to the namespace and include folder name - otherwise two data models will live in different namespaces, and they should be independent in the case.

I am not entirely sure I follow here. The first argument to `PODIO_GENERATE_DATAMODEL` is basically just the name of the folder into which the generated header files are put. It is not really related to the namespace that is used in the generated code. However it is usually a good idea to have a bit of consistency there. But you could for instance do something like this

```yaml
datatypes:
  fancy_namespace::FancyClass:
    Description: "Just an example"
    Author: "T. Madlener"
    Members:
      - int i
```

and then do 
```cmake
PODIO_GENERATE_DATAMODEL(datamodel datamodel.yaml headers sources)
```

However, it would get a bit awkward to use
```cpp
#include "datamodel/FancyClass.h"
using MyClass = fancy_namespace::FancyClass;
```
As you can see, the namespace and the include directory do not correspond to each other any longer. As far as I remember it is never enforced that the namespace that you define in the yaml file is the same as the name of the include directory name that you pass to `PODIO_GENERATE_DATAMODEL`.

So the only real problem is that you have when you pass the same name more than once to `PODIO_GENERATE_DATAMODEL` is that this will then pick up all the headers and all the source files from all previous generations, not just the current ones. So basically your first datamodel would only contain what you intended it to do, but your second datamodel would fully contain the first and the second one, and you would not even have to link the first against the second. This is something that could potentially be fixed with a bit of work around the generator and the surrounding CMake infrastructure.

But for your usecase it seems to be unnecessary, because you in any case already have `EDM4hep` to link against.

>  I'm not sure whether there is run time issue or not, but I believe it should run fine since they're plain c++ codes after the generation 

I haven't thought about this actually, but I think you are right, there should probably be no real difference.

**Comment by SanghyunKo** [comment]
I briefly tested your comment and it seems you're right. The problem is not `PODIO_GENERATE_DATAMODEL` itself but somewhat missing namespace modifier for `ConstFancyClass` when using `OneToOneRelations` or `OneToManyRelations` with a different namespace. For example, if I try following,
```yaml
datatypes :
  edm4dr::DRSimCalorimeterHit:
    Members:
      - int count // number of photons
    VectorMembers:
      - int timeStruct // time structure of detected photons
      - int wavlenSpectrum // wavelength spectrum of detected photons

    OneToOneRelations:
      - edm4hep::SimCalorimeterHit edm4hepSimCalorimeterHit // link to edm4hep::SimCalorimeterHit
    ExtraCode:
      includes:
        "#include <edm4hep/SimCalorimeterHit.h>\n #include <edm4hep/SimCalorimeterHitCollection.h>"
```

I get following error
```make
error: ‘ConstSimCalorimeterHit’ does not name a type; did you mean ‘ConstDRSimCalorimeterHit’?
 const ConstSimCalorimeterHit DRSimCalorimeterHit::getEdm4hepSimCalorimeterHit() const { ...
```
with the following problematic part in the generated source file `DRSimCalorimeterHit.cc`
```c++
const ConstSimCalorimeterHit DRSimCalorimeterHit::getEdm4hepSimCalorimeterHit() const {
  if (!m_obj->m_edm4hepSimCalorimeterHit) {
    return edm4hep::ConstSimCalorimeterHit(nullptr);
  }
  return edm4hep::ConstSimCalorimeterHit(*(m_obj->m_edm4hepSimCalorimeterHit));
}
```
Indeed, in the other files (including the header) the namespace is correctly specified
```c++
/// Access the link to edm4hep::SimCalorimeterHit
const ::edm4hep::ConstSimCalorimeterHit getEdm4hepSimCalorimeterHit() const;

/// Set the number of photons
void setCount(int value);


/// Set the link to edm4hep::SimCalorimeterHit
void setEdm4hepSimCalorimeterHit(::edm4hep::ConstSimCalorimeterHit value);
```



Anyway, I think the namespace issue is a relatively minor one, the concept should be applicable to any Podio data model in general, which may be quite useful to add some flexibility for Podio.

**Comment by tmadlener** [comment]
Thanks @SanghyunKo for checking this. That is definitely a bug in the generated code that you discovered there. I have opened #175 that should fix this.

**Comment by tmadlener** [comment]
Solved by #317.
